Avoid C/C++ math function ambiguity for explicit module builds on Windows

hjyamauchi opened the pull request:
This tweaks the existing logic that disables the import of the C++
math functions to avoid ambiguity with the C math functions by
handling the case where the module directory can be null in explicit
module builds.

For https://github.com/swiftlang/swift/issues/85606

compnerd commented on stdlib/public/Platform/tgmath.swift.gyb:
```diff
@@ -295,7 +295,11 @@ public func lgamma(_ x: ${T}) -> (${T}, Int) {
 @_transparent
 public func remquo(_ x: ${T}, _ y: ${T}) -> (${T}, Int) {
   var quo = Int32(0)
+#if !os(Windows)
```
Looking at that description, I think that this is the wrong solution. The problem is that `corecrt_math.h` does define a module, and the header being transitively imported in the headers by `CxxStdlib` results in the incorrect definition.

hjyamauchi replied:
My previous comment was not right. I think the headers are similar in macos. I updated the PR to explicitly call the C remquo. I think the conditions to detect `long double == double` case is captured by the `%  if T == 'Float80':` 6 lines above. I think this should work.

hjyamauchi commented:
@swift-ci please test

hjyamauchi committed:
Avoid C/C++ math function ambiguity for explicit module builds on Windows

This tweaks the existing logic that disables the import of the C++
math functions to avoid ambiguity with the C math functions by
handling the case where the module directory can be null in explicit
module builds.

For https://github.com/swiftlang/swift/issues/85606

hjyamauchi commented:
@swift-ci please test

compnerd commented:
I think that the check should be separated as it is unrelated to the check being modified. The `Directory` member may not be populated in some cases (but unrelated to explicit modules). This field represents the directory that the module is notionally built in. Since we inject the module maps via the VFS, the path may be unresolved, leaving the member not populated.

@ian-twilightcoder and @cachemeifyoucan might be able to shed more light on this.

hjyamauchi commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3814,7 +3785,8 @@ namespace {
             auto filename = file->getName();
-            if ((file->getDir() == owningModule->Directory) &&
+            // owningModule->Directory can be null in an explicit module build
+            if ((!owningModule->Directory || file->getDir() == owningModule->Directory) &&
```
Would the suggested change cause non-math functions to not be imported?

compnerd commented:
> I'm not all that familiar with the `Directory` member or how that gets put together. I'm not totally sure what the workaround is trying to work around either, neither cmath nor math.h are textual so I'm not sure what the issues are with those.

It seems that the field is optional and not always populated. Does it make sense to ignore the check for the directory? It seems that would open us up to cases where the header may be attributed to the wrong module.

Alternatively, should we bail early if the `Directory` member is not populated?

egorzhdan reviewed (approved):
LGTM, although I'm not too familiar with the testing approach here!

ian-twilightcoder commented:
> > I'm not all that familiar with the `Directory` member or how that gets put together. I'm not totally sure what the workaround is trying to work around either, neither cmath nor math.h are textual so I'm not sure what the issues are with those.
> 
> It seems that the field is optional and not always populated. Does it make sense to ignore the check for the directory? It seems that would open us up to cases where the header may be attributed to the wrong module.
> 
> Alternatively, should we bail early if the `Directory` member is not populated?

I don't really know, @Bigcheese @vsapsai @jansvoboda11 @benlangmuir do any of you know?

hjyamauchi commented:
@swift-ci please test linux platform

hjyamauchi commented:
@swift-ci please test linux platform

hjyamauchi commented:
@swift-ci please test

hjyamauchi commented:
If I remove the `Directory` check, I get following errors in [the Linux CI](https://ci.swift.org/job/swift-PR-Linux/25279/consoleText), meaning it unintentionally ends up over-hiding decls like `realpath` and `free` from `stdlib.h`.

Maybe this was masked by this `Directory` check? I am curious why https://github.com/swiftlang/swift/pull/71075 added this.

I think the original intention as in https://github.com/swiftlang/swift/pull/65437 was to hide the integer versions `abs` and `div` from `stdlib.h` in addition to the decls in `math.h` and the filename checks were *nested* inside the function name checks (`isAlternativeCStdlibFunctionFromTextualHeader`) but later they were somehow separated by https://github.com/swiftlang/swift/pull/66896.

I think removing the `Directory` check and but restricting the `stdlib.h` case to the variants of `abs` and `div` would be the way to go.

What do you think? @ian-twilightcoder @egorzhdan @zoecarver @benlangmuir @compnerd 

```
[2025-12-10T05:35:44.338Z] /home/build-user/swift/stdlib/public/RuntimeModule/Elf.swift:236:22: error: cannot find 'realpath' in scope
[2025-12-10T05:35:44.338Z]  234 | 
[2025-12-10T05:35:44.338Z]  235 | private func realPath(_ path: String) -> String? {
[2025-12-10T05:35:44.338Z]  236 |   guard let result = realpath(path, nil) else {
[2025-12-10T05:35:44.338Z]      |                      `- error: cannot find 'realpath' in scope
[2025-12-10T05:35:44.338Z]  237 |     return nil
[2025-12-10T05:35:44.338Z]  238 |   }
[2025-12-10T05:35:44.338Z] /home/build-user/swift/stdlib/public/RuntimeModule/Elf.swift:242:3: error: cannot find 'free' in scope
[2025-12-10T05:35:44.338Z]  240 |   let s = String(cString: result)
[2025-12-10T05:35:44.338Z]  241 | 
[2025-12-10T05:35:44.338Z]  242 |   free(result)
[2025-12-10T05:35:44.338Z]      |   `- error: cannot find 'free' in scope
[2025-12-10T05:35:44.338Z]  243 | 
[2025-12-10T05:35:44.338Z]  244 |   return s
[2025-12-10T05:35:44.338Z] 
[2025-12-10T05:35:44.338Z] /home/build-user/swift/stdlib/public/RuntimeModule/SymbolicatedBacktrace.swift:208:17: error: cannot find 'free' in scope
[2025-12-10T05:35:44.338Z] 206 |       if let demangled = _swift_backtrace_demangle(rawName, rawName.utf8.count,
[2025-12-10T05:35:44.338Z] 207 |                                                    nil, &length) {
[2025-12-10T05:35:44.338Z] 208 |         defer { free(demangled) }
[2025-12-10T05:35:44.338Z]     |                 `- error: cannot find 'free' in scope
[2025-12-10T05:35:44.338Z] 209 | 
[2025-12-10T05:35:44.338Z] 210 |         // length is the size of the buffer that was allocated, *not* the
```

hjyamauchi commented:
@swift-ci please test

compnerd reviewed (approved):
This makes a lot more sense to me. Thank you!
